src/backend/distributed/serialize_distributed_ddls.c

```diff
@@ -0,0 +1,122 @@
+/*-------------------------------------------------------------------------
+ *
+ * serialize_distributed_ddls.c
+ *
+ * This file contains functions for serializing distributed DDLs.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "miscadmin.h"
+
+#include "catalog/dependency.h"
+#include "storage/lock.h"
+
+#include "distributed/adaptive_executor.h"
+#include "distributed/argutils.h"
+#include "distributed/deparse_shard_query.h"
+#include "distributed/resource_lock.h"
+#include "distributed/serialize_distributed_ddls.h"
+
+
+PG_FUNCTION_INFO_V1(citus_internal_acquire_citus_advisory_object_class_lock);
+
+
+static char * AcquireCitusAdvisoryObjectClassLockCommand(ObjectClass objectClass);
+static void AcquireCitusAdvisoryObjectClassLock(ObjectClass objectClass);
+
+
+/*
+ * citus_internal_acquire_citus_advisory_object_class_lock is an internal UDF
+ * to call AcquireCitusAdvisoryObjectClassLock().
+ */
+Datum
+citus_internal_acquire_citus_advisory_object_class_lock(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	PG_ENSURE_ARGNOTNULL(0, "object_class");
+	ObjectClass objectClass = PG_GETARG_INT32(0);
+
```

Comment:
Let's put some security mechanism in place to check that not just any user can block create database DDL. I think the best would be to check if the user has CREATEDB privileges. If they do, then they are allowed to acquire the lock.

Reply:
Thinking loudly ..

We're acquiring this lock not only for create database but also for alter / drop database commands to prevent the deadlocks that could happen, say, when issuing "create database X" & "alter database rename Y to X" concurrently.

For create database, it makes sense to ensure CREATEDB privilage. However for ALTER / DROP DATABASE, it would probably make more sense to check db ownership.

However, I'm not sure how we can do that in this UDF as it doesn't know about what type of command we're executing.